### PR TITLE
[SPARK-30842][SQL] Adjust abstraction structure for join operators

### PR DESCRIPTION
**Pull request by Eric5553**
### What changes were proposed in this pull request?
Currently the join operators are not well abstracted, since there are lot of common logic. A trait can be created for easier pattern matching and other future handiness. This is a follow-up PR based on comment 
https://github.com/apache/spark/pull/27509#discussion_r379613391 .

This PR refined from the following aspects:
1. Refined structure of all physical join operators
2. Add missing joinType field for CartesianProductExec operator
3. Refined codes related to Explain Formatted

The EXPLAIN FORMATTED changes are
1. Converge all join operator `verboseStringWithOperatorId` implementations to `BaseJoinExec`. Join condition displayed, and join keys displayed if it’s not empty.
2. `#1` will add Join condition to `BroadcastNestedLoopJoinExec`.
3. `#1` will **NOT** affect `CartesianProductExec`,`SortMergeJoin` and `HashJoin`s, since they already got there override implementation before.
4. Converge all join operator `simpleStringWithNodeId` to `BaseJoinExec`, which will enhance the one line description for `CartesianProductExec` with `JoinType` added.
5. Override `simpleStringWithNodeId` in `BroadcastNestedLoopJoinExec` to show `BuildSide`, which was only done for `HashJoin`s before.

### Why are the changes needed?
Make the code consistent with other operators and for future handiness of join operators.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by Eric5553** [comment]
cc @maryannxue @gatorsmile @cloud-fan @maropu

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  override def verboseStringWithOperatorId(): String = {
+    val joinCondStr = if (condition.isDefined) {
+      s"${condition.get}"
+    } else "None"
+    s"""
+       |(${ExplainUtils.getOpId(this)}) $nodeName ${ExplainUtils.getCodegenId(this)}
+       |Join condition : ${joinCondStr}
+     """.stripMargin
+  }
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: We need this trait for join physical plans? There seems to be less common method between them...

Comment (Eric5553) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: Yea, when implementing the suggested follow-up, I found only `joinType` and `conditions` can be pull up to base trait. And even the `joinType` not found in `CartesianProductExec`. 

@maryannxue Does this PR meets your suggestion? Just let me know if I misunderstood anything, thanks! :-)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/CartesianProductExec.scala [hunk 2]
```diff
@@ -60,23 +61,13 @@ class UnsafeCartesianRDD(
 case class CartesianProductExec(
     left: SparkPlan,
     right: SparkPlan,
-    condition: Option[Expression]) extends BinaryExecNode {
+    joinType: JoinType,
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/CartesianProductExec.scala]: CartesianProduct is always an inner join.

Comment (Eric5553) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/CartesianProductExec.scala]: Thanks for review!  Addressed in 4397ce88a99b4777067e7a636f12e22612b04be5.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -286,7 +286,8 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
         def createCartesianProduct() = {
           if (joinType.isInstanceOf[InnerLike]) {
-            Some(Seq(joins.CartesianProductExec(planLater(left), planLater(right), condition)))
+            Some(Seq(joins.CartesianProductExec(
+              planLater(left), planLater(right), condition)))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -367,7 +368,8 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: maybe we can also put left/right join keys here. The join keys are Nil in some sub-classes like `CartesianProductExec`.

Comment (maryannxue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: Thank you @Eric5553! this looks good. Adding equi-join keys sounds good to me too.

Comment (Eric5553) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: @cloud-fan @maryannxue left/right join keys are also added to `BaseJoinExec`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 5]
```diff
@@ -43,6 +46,11 @@ case class BroadcastNestedLoopJoinExec(
     case BuildLeft => (right, left)
   }
 
+  override def simpleStringWithNodeId(): String = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: do we need to override it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    s"$nodeName $joinType ($opId)".trim
+  }
+
+  override def verboseStringWithOperatorId(): String = {
+    val joinCondStr = if (condition.isDefined) {
+      s"${condition.get}"
+    } else "None"
+    s"""
+       |(${ExplainUtils.getOpId(this)}) $nodeName ${ExplainUtils.getCodegenId(this)}
+       |${ExplainUtils.generateFieldString("Join condition", joinCondStr)}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: can we include join keys here? Then we can remove the `verboseStringWithOperatorId` methods in join sub-classes.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: And we should print nothing if join keys are empty, instead of `[]`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 5 above)
Comment (Eric5553) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: It's to show the `BuildSide` of `BroadcastNestedLoopJoinExec`, I think we missed it here. We don't need to move`BuildSide` to `BaseJoinExec`, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala (hunk 6 above)
Comment (Eric5553) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BaseJoinExec.scala]: Sure, will update. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Yea we don't need

**Comment by cloud-fan** [comment]
LGTM. @Eric5553 can you list the EXPLAIN output changes introduced by this PR? Thanks!

**Comment by Eric5553** [comment]
@cloud-fan Sure
The `EXPLAIN FORMATTED` changes are:
1. Converge all join operator `verboseStringWithOperatorId` implementations to `BaseJoinExec`. Join condition displayed, and join keys displayed if it’s not empty.
2. `#1` will add Join condition to `BroadcastNestedLoopJoinExec`.
3. `#1` will **NOT** affect `CartesianProductExec`,`SortMergeJoin` and `HashJoin`s, since they already got there override implementation before.
4. Converge all join operator `simpleStringWithNodeId` to `BaseJoinExec`, which will enhance the one line description for `CartesianProductExec` with `JoinType` added.
5. Override `simpleStringWithNodeId` in `BroadcastNestedLoopJoinExec` to show `BuildSide`, which was only done for `HashJoin`s before.

Also updated in PR descriptor.

**Comment by cloud-fan** [comment]
@Eric5553 can you fix the conflicts? thanks!

**Comment by Eric5553** [comment]
retest this please

**Comment by Eric5553** [comment]
@cloud-fan Sure, the conflicts have been resolved. :-)

**Comment by cloud-fan** [comment]
thanks, merging to master!
